library/X509/CertificateUtils.php

```diff
@@ -504,18 +496,28 @@ public static function verifyCertificates(Connection $db)
 
                 preg_match('/^error \d+ at \d+ depth lookup:(.+)$/m', $output, $match);
 
-                if (!empty($match)) {
+                if (! empty($match)) {
                     $set = ['invalid_reason' => trim($match[1])];
                 } else {
                     $set = ['valid' => 'y', 'invalid_reason' => null];
                 }
 
                 // TODO: https://github.com/Icinga/ipl-orm/pull/78
-                $db->update(
-                    'x509_certificate_chain',
-                    $set,
-                    ['id = ?' => $chain->id]
-                );
+                $db->update('x509_certificate_chain', $set, ['id = ?' => $chainId]);
+            };
+
+            foreach ($certs as $cert) {
```

Comment:
I'd like to make this a library function, e.g.

```
/**
 * Yield sets of items from an iterator grouped by a specific criterion gathered from a callback
 *
 * @param Iterator $iterator
 * @param callable $groupBy
 *
 * @return Generator
 */
function yield_groups(Iterator $iterator, callable $groupBy): Generator
{
    if (! $iterator->valid()) {
        return;
    }

    $attribute = $groupBy($iterator->current(), $iterator->key());
    $group = [$iterator->current()];
//    $group = [$iterator->key() => $iterator->current()];

    $iterator->next();
    for (; $iterator->valid(); $iterator->next()) {
        $a = $groupBy($iterator->current(), $iterator->key());
        if ($a !== $attribute) {
            yield $attribute => $group;

            $group = [];
            $attribute = $a;
        }
        
        $group[] = $iterator->current();
//        $group[$iterator->key()] = $iterator->current();
    }

    if (! empty($group)) {
        yield $attribute => $group;
    }
}
```

What do you think? And @nilmerg what do you think?

Reply:
When the values to be grouped, i.e. `$group[] = $iterator->current()` could also be specified by a callback, it would work. Such a grouping function would also be used for Sla queries in Icinga DB Web.

Reply:
Sure. Though, why not foreach?